File: graph/dfs.cpp
#include <iostream>
#include <vector>

#define MAX 100005

using namespace std;

vector<int> adj[MAX];
bool visited[MAX];

void addEdge(int a, int b)
{
    adj[a].push_back(b);
    adj[b].push_back(a);
}

void DFS(int u)
{
    visited[u] = true;
    cout << u << ' ';

    for (auto i: adj[u])
    {
        if (!visited[i])
        {
            DFS(i);
        }
    }
}

int main()
{
    int v, e;

    cout << "Enter the number of vertices: ";
    cin >> v;
    cout << "Enter the number of edges: ";
    cin >> e;

    cout << "Enter the edge details:\n";

    int a, b;

    for (int i = 0; i < v; i++)
    {
        visited[i] = false;
        adj[i].clear();
    }

    while (e--)
    {
        cin >> a >> b;
        addEdge(a, b);
    }

    // Depth-First-Search starting from vertex 0.
    DFS(0);

    return 0;
}

File: graph/dfs.py
def dfs(u):
    visited[u] = True

    print(u, end=' ')

    for neighbour in adj[u]:
        if not visited[neighbour]:
            dfs(neighbour)


v, e = list(map(int, input().split()))

adj = [[] for _ in range(v)]
visited = [False] * (v)

for i in range(e):
    a, b = list(map(int, input().split())) 
    adj[a].append(b)
    adj[b].append(a)

    # Assuming the graph is undirected.

for i in range(v):
    if not visited[i]:
        dfs(i)
